File: jotupix_demo_ios/JotupixDemo/Jotupix/Program/JProgramContent.m
// JProgramContent.m
#import "JProgramContent.h"
#import "JByteWriter.h"

@interface JProgramContent ()
@property (nonatomic, strong) NSMutableArray<JContentBase *> *contentList;   /// Internal storage for content objects
@end

@implementation JProgramContent

/// Initializes the container with an empty content list.
- (instancetype)init {
    if (self = [super init]) {
        _contentList = [NSMutableArray array];
    }
    return self;
}

#pragma mark - API

/// Adds a content object to the container.
/// Ignores nil values.
- (void)addContent:(JContentBase *)content {
    if (content) {
        [self.contentList addObject:content];
    }
}

/// Clears all content objects from the container.
- (void)clear {
    [self.contentList removeAllObjects];
}

/// Serializes all content items into a single data block following the protocol layout.
///
/// Layout:
///   [0..7]  : 8 reserved bytes
///   [8]     : number of content blocks
///   [9]     : reserved byte
///   [...]   : serialized content blocks (in the same order added)
///
/// Each content object is responsible for producing its own formatted data.
- (NSData *)get {

    JByteWriter *proContent = [[JByteWriter alloc] init];

    // reserved (8 bytes)
    [proContent put_repeat:0 count:8];

    // number of content blocks
    [proContent put_u8:(uint8_t)self.contentList.count];

    // reserved
    [proContent put_u8:0];

    // write each content block
    for (JContentBase *content in self.contentList) {
        NSData *block = [content get];
        if (block) {
            [proContent put_bytes:block];
        }
    }

    // return NSData, NOT a raw pointer
    return [NSData dataWithBytes:proContent.buffer length:proContent.size];
}

@end

File: jotupix_demo_windows/jotupix/program/JProgramContent.cpp
#include "JProgramContent.h"
#include "JTextContent.h"
#include "JByteWriter.h"
#include "JTextDiyColorContent.h"
#include "JTextFullColorContent.h"
#include "JGifContent.h"

void JProgramContent::Add(std::shared_ptr<JContentBase> content)
{
	m_contentData.push_back(content);
}

void JProgramContent::Clear()
{
	m_contentData.clear();
}

std::vector<uint8_t> JProgramContent::Get()
{
	JByteWriter proContent;

	proContent.put_repeat(0, 8); // reserved
	proContent.put_u8(m_contentData.size());
	proContent.put_u8(0); // reserved
	
	for (const auto& content : m_contentData)
	{
		proContent.put_bytes(content->Get());
	}

	return proContent.buffer;
}
